base/applications/cmdutils/fc/fc.c

```diff
@@ -0,0 +1,498 @@
+                                        HANDLE hMapping2, size_t cb2)
+{
+    FCRET ret;
+    BOOL fIgnoreCase = !!(pFC->dwFlags & FLAG_C);
+    DWORD dwCmpFlags = (fIgnoreCase ? NORM_IGNORECASE : 0);
+    LPSTR psz1, psz2;
+    size_t ib = 0;
+
+    do
+    {
+        psz1 = MapViewOfFile(hMapping1, FILE_MAP_READ, HILONG(ib), LOLONG(ib), (DWORD)cb1);
+        psz2 = MapViewOfFile(hMapping2, FILE_MAP_READ, HILONG(ib), LOLONG(ib), (DWORD)cb2);
+        if (!psz1 || !psz2)
+        {
+            ret = OutOfMemory();
+            break;
+        }
+        if (CompareStringA(0, dwCmpFlags, psz1, (INT)cb1, psz2, (INT)cb2) == CSTR_EQUAL)
+        {
+            ret = NoDifference();
+            break;
+        }
+        // TODO: compare each lines
+        // TODO: large file support
+        ret = Different(pFC->file1, pFC->file2);
+    } while (0);
+
+    UnmapViewOfFile(psz1);
+    UnmapViewOfFile(psz2);
+    return ret;
+}
+
+static FCRET TextFileCompare(const FILECOMPARE *pFC)
+{
+    FCRET ret;
+    HANDLE hFile1, hFile2, hMapping1 = NULL, hMapping2 = NULL;
+    size_t cb1, cb2;
+    BOOL fUnicode = !!(pFC->dwFlags & FLAG_U);
+
+    hFile1 = DoOpenFileForInput(pFC->file1);
+    if (hFile1 == INVALID_HANDLE_VALUE)
+        return FCRET_CANT_FIND;
+    hFile2 = DoOpenFileForInput(pFC->file2);
+    if (hFile2 == INVALID_HANDLE_VALUE)
+    {
+        CloseHandle(hFile1);
+        return FCRET_CANT_FIND;
+    }
+
+    do
+    {
+        if (lstrcmpiW(pFC->file1, pFC->file2) == 0)
+        {
+            ret = NoDifference();
+            break;
+        }
+        if (!GetFileSizeDx(hFile1, &cb1, FALSE))
+        {
+            ret = CannotRead(pFC->file1);
+            break;
+        }
+        if (!GetFileSizeDx(hFile2, &cb2, FALSE))
+        {
+            ret = CannotRead(pFC->file2);
+            break;
+        }
+        if (cb1 == 0 && cb2 == 0)
+        {
+            ret = NoDifference();
+            break;
+        }
+        hMapping1 = CreateFileMappingW(hFile1, NULL, PAGE_READONLY,
+                                       HILONG(cb1), LOLONG(cb1), NULL);
+        if (hMapping1 == NULL)
+        {
+            ret = CannotRead(pFC->file1);
+            break;
+        }
+        hMapping2 = CreateFileMappingW(hFile2, NULL, PAGE_READONLY,
+                                       HILONG(cb2), LOLONG(cb2), NULL);
+        if (hMapping2 == NULL)
+        {
+            ret = CannotRead(pFC->file2);
+            break;
+        }
+
+        if (fUnicode)
+            ret = UnicodeTextCompare(pFC, hMapping1, cb1, hMapping2, cb2);
+        else
+            ret = AnsiTextCompare(pFC, hMapping1, cb1, hMapping2, cb2);
+    } while (0);
+
+    CloseHandle(hMapping1);
+    CloseHandle(hMapping2);
+    CloseHandle(hFile1);
+    CloseHandle(hFile2);
+    return ret;
+}
+
+static BOOL IsBinaryExt(LPCWSTR filename)
+{
+    static const LPCWSTR s_dotexts[] =
+    {
+        L".exe", L".com", L".sys", L".obj", L".lib", L".bin"
+    };
```

Comment:
Isn't there a more robust way to get this list? It seems pretty arbitrary (the .obj,, .lib and .bin parts in particular)

Reply:
Registry?

Reply:
or BOM?

Reply:
https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/fc

> /l : Compares the files in ASCII mode, line-by-line, and attempts to resynchronize the files after finding a mismatch. This is the default mode for comparing files, except files with the following file extensions: `.exe`, `.com`, `.sys`, `.obj`, `.lib`, or `.bin`.

Reply:
Ok then, this should be mentioned in a comment in the code.

Reply:
This is by design.

Reply:
OK.

Reply:
It appears that Windows Server 2003 hard codes these extensions inside its fc.exe file as can be see here:

![fc_exe_exts](https://user-images.githubusercontent.com/32620577/115977889-8eab9980-a541-11eb-8598-85d72c31be83.png)